vault/request_handling.go

```diff
@@ -0,0 +1,425 @@
+package vault
+
+import (
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+var (
+	// Value for memoizing whether cubbyhole is mounted, e.g. if we are in
+	// normal operation and not test mode
+	cubbyholeMounted bool
+
+	// mutex to ensure the same
+	cubbyholeMountedMutex sync.Mutex
+)
+
+// HandleRequest is used to handle a new incoming request
+func (c *Core) HandleRequest(req *logical.Request) (resp *logical.Response, err error) {
+	c.stateLock.RLock()
+	defer c.stateLock.RUnlock()
+	if c.sealed {
+		return nil, ErrSealed
+	}
+	if c.standby {
+		return nil, ErrStandby
+	}
+
+	// Allowing writing to a path ending in / makes it extremely difficult to
+	// understand user intent for the filesystem-like backends (generic,
+	// cubbyhole) -- did they want a key named foo/ or did they want to write
+	// to a directory foo/ with no (or forgotten) key, or...? It also affects
+	// lookup, because paths ending in / are considered prefixes by some
+	// backends. Basically, it's all just terrible, so don't allow it.
+	if strings.HasSuffix(req.Path, "/") &&
+		(req.Operation == logical.UpdateOperation ||
+			req.Operation == logical.CreateOperation) {
+		return logical.ErrorResponse("cannot write to a path ending in '/'"), nil
+	}
+
+	var auth *logical.Auth
+	if c.router.LoginPath(req.Path) {
+		resp, auth, err = c.handleLoginRequest(req)
+	} else {
+		resp, auth, err = c.handleRequest(req)
+	}
+
+	// Ensure we don't leak internal data
+	if resp != nil {
+		if resp.Secret != nil {
+			resp.Secret.InternalData = nil
+		}
+		if resp.Auth != nil {
+			resp.Auth.InternalData = nil
+		}
+	}
+
+	// In order to wrap, we need cubbyhole to be mounted, so we ensure that
+	// cubbyhole is actually mounted, as it may not be during tests. We memoize
+	// a true response, since cubbyhole cannot be mounted or unmounted during
+	// normal operation.
+	if !cubbyholeMounted {
+		cubbyholeMountedMutex.Lock()
+		// Ensure it wasn't changed by another goroutine
+		if !cubbyholeMounted {
+			if c.router.MatchingMount("cubbyhole/") != "" {
+				cubbyholeMounted = true
+			}
+		}
+		cubbyholeMountedMutex.Unlock()
+	}
+
+	// We are wrapping if there is anything to wrap (not a nil response) and a
+	// TTL was specified for the token, plus if cubbyhole is mounted (which
+	// will be the case normally)
+	wrapping := cubbyholeMounted && resp != nil && resp.WrapInfo != nil && resp.WrapInfo.TTL != 0
```

Comment:
I would change this around, if cubbyhole is disabled and wrapping is requested, the request should just be failed immediately.